include/swift/SwiftRemoteMirror/MemoryReaderInterface.h

```diff
@@ -32,66 +32,70 @@ extern "C" {
+/// would go beyond the section boundary, a length of 0 should be
+/// returned.
+///
+/// \param address the address in the target address space
+/// \returns The length of the string or 0 if the scan was unsuccessful.
 typedef uint64_t (*GetStringLengthFunction)(void *reader_context,
                                             swift_addr_t address);
+
+/// Get the address of a symbol in the target address space.
+///
+/// \returns true if the lookup was successful.
 typedef swift_addr_t (*GetSymbolAddressFunction)(void *reader_context,
                                                  const char *name,
                                                  uint64_t name_length);
 
-typedef struct MemoryReaderImpl {
-  /// An opaque context that the implementor can specify to
-  /// be passed to each of the APIs below.
-  void *reader_context;
-
-  /// Get the size in bytes of the target's pointer type.
-  PointerSizeFunction getPointerSize;
-
-  /// Get the size in bytes of the target's size type.
-  SizeSizeFunction getSizeSize;
-
-  /// Free memory returned from readBytes. May be NULL if memory never needs to be freed.
-  FreeBytesFunction free;
-
-  // FIXME: -Wdocumentation complains about \param and \returns on function pointers.
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdocumentation"
-
-  /// Read a sequence of bytes at an address in the target.
-  ///
-  /// \param address the address in the target address space
-  /// \param size the number of bytes to read
-  /// \param outFreeContext on return, an arbitrary context pointer that the caller will
-  ///                       pass to the free function
-  /// \returns A pointer to the requested memory, or NULL if the memory could not be read.
-  ///          The caller must invoke the free function on the returned pointer once it's
-  ///          done using the memory.
-  ReadBytesFunction readBytes;
-
-  /// Get the string length at the given address.
-  ///
-  /// This scan always occurs in a read-only data section. If the scan
-  /// would go beyond the section boundary, a length of 0 should be
-  /// returned.
-  ///
-  /// \param address the address in the target address space
-  /// \returns The length of the string or 0 if the scan was unsuccessful.
-  GetStringLengthFunction getStringLength;
-
-  /// Get the address of a symbol in the target address space.
-  ///
-  /// \returns true if the lookup was successful.
-  GetSymbolAddressFunction getSymbolAddress;
-
-#pragma clang diagnostic pop
-
-} MemoryReaderImpl;
+typedef enum {
+  /// The query should ignore inBuffer, and treat outBuffer as uint8_t* which
+  /// should be populated with the size of an ordinary pointer in the remote
+  /// process, in bytes.
+  PointerSize,
+
+  /// The query should ignore inBuffer, and treat outBuffer as uint8_t* which
+  /// should be populated with the size of size_t in the remote process, in
+  /// bytes.
+  SizeSize,
+} DataLayoutQueryType;
```

Comment:
Since this is a C interface and we can't use `enum class`, I think you should prefix these enumerators.

I would suggest `DLQ_GetPointerSize` and `DLQ_GetSizeSize`.